examples/bip-340/src/bip-340.rs

```diff
@@ -0,0 +1,169 @@
+use hacspec_lib::*;
+use hacspec_sha256::*;
+
+// TODO: Need to add some explanation what this does
+public_nat_mod!(
+    type_name: FieldElement,
+    type_of_canvas: FieldCanvas,
+    bit_size_of_field: 256,
+    modulo_value: "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEFFFFFC2F"
+);
+public_nat_mod!(
+    type_name: Scalar,
+    type_of_canvas: ScalarCanvas,
+    bit_size_of_field: 256,
+    modulo_value: "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141"
+);
+public_nat_mod!(
+    type_name: BigInteger,
+    type_of_canvas: BigIntegerCanvas,
+    bit_size_of_field: 256,
+    modulo_value: "8000000000000000000000000000000080000000000000000000000000000000"
+);
+
+type FinitePoint = (FieldElement, FieldElement);
+
+// TODO: Point at infinity is represented as None. There should be better ways
+// to do this.
+type Point = Option<FinitePoint>;
+
+bytes!(SerializedScalar, 32);
+pub type SecretKey = SerializedScalar;
+bytes!(PublicKey, 32);
+bytes!(Message, 32);
+bytes!(AuxRand, 32);
+bytes!(Signature, 64);
+// TODO: is this really necessary just to define G?
+bytes!(RawPoint, 64);
+
+// TODO: This shouldn't be a secret array, since there's nothing secret about it
+#[rustfmt::skip]
+const G: RawPoint = RawPoint(secret_array!(
+    U8,
+    [
+        0x79u8, 0xBEu8, 0x66u8, 0x7Eu8, 0xF9u8, 0xDCu8, 0xBBu8, 0xACu8, 0x55u8, 0xA0u8, 0x62u8, 0x95u8, 0xCEu8, 0x87u8, 0x0Bu8, 0x07u8, 0x02u8, 0x9Bu8, 0xFCu8, 0xDBu8, 0x2Du8, 0xCEu8, 0x28u8, 0xD9u8, 0x59u8, 0xF2u8, 0x81u8, 0x5Bu8, 0x16u8, 0xF8u8, 0x17u8, 0x98u8,
+        0x48u8, 0x3Au8, 0xDAu8, 0x77u8, 0x26u8, 0xA3u8, 0xC4u8, 0x65u8, 0x5Du8, 0xA4u8, 0xFBu8, 0xFCu8, 0x0Eu8, 0x11u8, 0x08u8, 0xA8u8, 0xFDu8, 0x17u8, 0xB4u8, 0x48u8, 0xA6u8, 0x85u8, 0x54u8, 0x19u8, 0x9Cu8, 0x47u8, 0xD0u8, 0x8Fu8, 0xFBu8, 0x10u8, 0xD4u8, 0xB8u8
+
+    ]
+));
+
+pub fn tagged_hash(tag: String, msg: &ByteSeq) -> Sha256Digest {
+    let tag_hash = sha256(&ByteSeq::from_string(tag)).to_be_bytes();
+    sha256(&tag_hash.concat(&tag_hash).concat(msg))
+}
+
+fn x(p: FinitePoint) -> FieldElement {
+    let (x, _) = p;
+    x
+}
+
+fn y(p: FinitePoint) -> FieldElement {
+    let (_, y) = p;
+    y
+}
+
+fn has_even_y(p: FinitePoint) -> bool {
+    y(p) % FieldElement::TWO() == FieldElement::ZERO()
+}
+
+fn compute_lam(p1_: FinitePoint, p2_: FinitePoint) -> FieldElement {
+    let three = FieldElement::TWO() + FieldElement::ONE();
+    if p1_ != p2_ {
+        (y(p2_) - y(p1_)) * (x(p2_) - x(p1_)).pow_self(FieldElement::ZERO() - FieldElement::TWO())
+    } else {
+        three
+            * x(p1_)
+            * x(p1_)
+            * (FieldElement::TWO() * y(p1_)).pow_self(FieldElement::ZERO() - FieldElement::TWO())
+    }
+}
+
+fn point_add(p1: Point, p2: Point) -> Point {
+    // TODO: I tried writing this as an if/else if/else case distinction
+    // (without res) but hacspec doesn't let me do that.
```

Comment:
Yeah... this is the unfortunately still open #85

Reply:
Ah, thx, good to know.